# Introduction
This project focuses on predicting the winner of football matches in the English Premier League (EPL) from seasons 2020-21 and 2021-22. Through using the data provided [here]("https://fbref.com/en/comps/9/Premier-League-Stats") and a **random forest classifier** algorithm from the sickitlearn library, the code aims at predicting match results. It's worthy of mention that the 1st version of the code focuses on weather a team has won a game or not rather than trying to predict weather the result was a tie, lost or win.

## Phase 0: Taking a Look at the Provided Dataframe

In [1]:
import pandas as pd

In [2]:
matches = pd.read_csv('matches.csv', index_col=0)
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [3]:
matches.shape

(1389, 27)

The EPL has 20 teams each seasons. So the number of matches played by each team is 38 (Remember that each team plays 2 games with the other teams, one a home match the other, an away match). 

Since the dataset contains data related to 2 EPL seasons, the total number of matches is 

$$38 \times 20 \times 2 = 1520$$

while the data provided contains only 1389 records.

So there shall be an investigation to see what data we are missing

In [4]:
matches['team'].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

We know that in each season, **3** teams get relegated to English Football League Championship, so there shall be **6** different teams with fewer than 70 matches in our dataset, but the output above contains 7 teams with the mentioned conditions.
After looking at the stats of these 7 teams, we can see that **Liverpool** stats look a bit different as shown below

In [5]:
matches[matches['team'] == 'Liverpool'].head(7)

date   time            comp        round  day venue result   gf  \
1   2020-09-12  17:30  Premier League  Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League  Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League  Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League  Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League  Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League  Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League  Matchweek 7  Sat  Home      W  2.0   

     ga       opponent  ...  match report  notes    sh  sot  dist   fk   pk  \
1   3.0   Leeds United  ...  Match Report    NaN  20.0  4.0  17.0  0.0  2.0   
2   0.0        Chelsea  ...  Match Report    NaN  17.0  5.0  17.7  1.0  0.0   
4   1.0        Arsenal  ...  Match Report    NaN  21.0  9.0  16.8  0.0  0.0   
6   7.0    Aston Villa  ...  Match Report    NaN  14.0  8.0  15.8  1.0  0.0   
7   2.0        Everton  ...  Match Report    NaN  22.0  8.0  15.0  1.0  0.0   
9   1.0  Sheffield Utd  ...  Match Report    NaN  17.0  5.0  18.2  1.0  0.0   
11  1.0       West Ham  ...  Match Report    NaN   8.0  2.0  18.6  1.0  1.0   

   pkatt  season       team  
1    2.0    2021  Liverpool  
2    0.0    2021  Liverpool  
4    0.0    2021  Liverpool  
6    0.0    2021  Liverpool  
7    0.0    2021  Liverpool  
9    0.0    2021  Liverpool  
11   1.0    2021  Liverpool  

[7 rows x 27 columns]

Taking a close look at the table above, wa can see that the table is missing data for the 2021-22 season for Liverpool. (Thus, containing the stats for a single season only)

The reason is that the provided data was being collected while the 2021-22 season was still in it's final weeks and not finished. (Looking at the results below we can confirm that)

In [6]:
matches['round'].value_counts()

Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: round, dtype: int64

## Phase 1: Data Cleaning
In order to use the data provided for our project, it's necessary to clean up the dataframe.

In [7]:
matches.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

An important aspect of data in machine learning projects is the type of data being used, because most ML algorithms **can't** be performed on any type of data other than integers.

Taking a look at the table above, we see `object` type in our records while ML algorithms cannot work with this data type, so we shall find a way to use these columns.

#### 1. Converting `date` column to `date-time` type

In [8]:
matches['date'] = pd.to_datetime(matches['date'])

In [9]:
matches.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

Now the `date` column contains records of type `datetime64[ns]` which is desirable.

#### 2. Creating predictors for ML

In [10]:
matches['venue_code'] = matches['venue'].astype('category').cat.codes

In [11]:
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  notes    sh   sot  dist   fk   pk pkatt season  \
1       Tottenham  ...    NaN  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2    Norwich City  ...    NaN  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3         Arsenal  ...    NaN  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4  Leicester City  ...    NaN  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6     Southampton  ...    NaN  16.0   1.0  15.7  1.0  0.0   0.0   2022   

              team  venue_code  
1  Manchester City           0  
2  Manchester City           1  
3  Manchester City           1  
4  Manchester City           0  
6  Manchester City           1  

[5 rows x 28 columns]

In [12]:
matches['opponent_code'] = matches['opponent'].astype('category').cat.codes

In [13]:
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...    sh   sot  dist   fk   pk pkatt season  \
1       Tottenham  ...  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2    Norwich City  ...  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3         Arsenal  ...  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4  Leicester City  ...  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6     Southampton  ...  16.0   1.0  15.7  1.0  0.0   0.0   2022   

              team  venue_code  opponent_code  
1  Manchester City           0             18  
2  Manchester City           1             15  
3  Manchester City           1              0  
4  Manchester City           0             10  
6  Manchester City           1             17  

[5 rows x 29 columns]

In [14]:
matches['hour'] = matches['time'].str.replace(':.+', '', regex=True).astype('int')

In [15]:
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   sot  dist   fk   pk pkatt season             team  \
1       Tottenham  ...   4.0  16.9  1.0  0.0   0.0   2022  Manchester City   
2    Norwich City  ...   4.0  17.3  1.0  0.0   0.0   2022  Manchester City   
3         Arsenal  ...  10.0  14.3  0.0  0.0   0.0   2022  Manchester City   
4  Leicester City  ...   8.0  14.0  0.0  0.0   0.0   2022  Manchester City   
6     Southampton  ...   1.0  15.7  1.0  0.0   0.0   2022  Manchester City   

  venue_code  opponent_code  hour  
1          0             18    16  
2          1             15    15  
3          1              0    12  
4          0             10    15  
6          1             17    15  

[5 rows x 30 columns]

In [16]:
matches['day_code'] = matches['date'].dt.dayofweek

In [17]:
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  dist   fk   pk  pkatt season             team  \
1       Tottenham  ...  16.9  1.0  0.0    0.0   2022  Manchester City   
2    Norwich City  ...  17.3  1.0  0.0    0.0   2022  Manchester City   
3         Arsenal  ...  14.3  0.0  0.0    0.0   2022  Manchester City   
4  Leicester City  ...  14.0  0.0  0.0    0.0   2022  Manchester City   
6     Southampton  ...  15.7  1.0  0.0    0.0   2022  Manchester City   

  venue_code opponent_code  hour  day_code  
1          0            18    16         6  
2          1            15    15         5  
3          1             0    12         5  
4          0            10    15         5  
6          1            17    15         5  

[5 rows x 31 columns]

#### 3. Set Target Value 

In [18]:
matches['target'] = (matches['result'] == 'W').astype('int')

In [19]:
matches.head(5)

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   fk   pk  pkatt  season             team venue_code  \
1       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City          0   
2    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City          1   
3         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City          1   
4  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City          0   
6     Southampton  ...  1.0  0.0    0.0    2022  Manchester City          1   

  opponent_code hour  day_code  target  
1            18   16         6       0  
2            15   15         5       1  
3             0   12         5       1  
4            10   15         5       1  
6            17   15         5       0  

[5 rows x 32 columns]

## Phase 2: Designing the ML Model

We will use a random forest classifier due to the fact that the values being predicted are not linear.

In [20]:
from sklearn.ensemble import RandomForestClassifier

In [21]:
rf = RandomForestClassifier(n_estimators=50, 
                            min_samples_split=10, 
                            random_state=1)

**Note**

It's important that the test data comes after the train data in their respective time order while working with a time series dataset, otherwise the prediction wouldn't be valid. (Because the model would be tested on data records relevant to time stamps before what it has been trained on which is not correct and logical!)

In [22]:
train = matches[matches['date'] < '2022-01-01'] # take all matched before 2022
test = matches[matches['date'] > '2022-01-01'] # take all matched before 2022

In [23]:
predictors = ['venue_code', 'opponent_code', 'hour', 'day_code']

In [24]:
rf.fit(train[predictors], train['target'])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [25]:
preds = rf.predict(test[predictors])

## Phase 3: Determine accuracy of the model

In [26]:
from sklearn.metrics import accuracy_score

In [27]:
acc = accuracy_score(test['target'], preds)

In [28]:
acc 

0.6123188405797102

The accuracy is 61 percent at the moment. In the following pieces of code, we will work on improving the predictions.

## Phase 4: Processing the data to improve predictions

In [29]:
combined = pd.DataFrame(dict(actual=test['target'], prediction=preds))

In [30]:
pd.crosstab(index=combined['actual'], columns=combined['prediction'])

prediction    0   1
actual             
0           141  31
1            76  28

In [31]:
from sklearn.metrics import precision_score

In [32]:
precision_score(test['target'], preds)

0.4745762711864407

Creating more predictors to improve accuracy model
through computing **rolling averages** for team performance

In [33]:
grouped_matches = matches.groupby('team')  # create 1 data frame for each team in our data

In [34]:
group = grouped_matches.get_group("Manchester City")

In [35]:
group  # get all matches that mancity played in

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
54 2021-05-01  12:30  Premier League  Matchweek 34  Sat  Away      W  2.0   
56 2021-05-08  17:30  Premier League  Matchweek 35  Sat  Home      L  1.0   
57 2021-05-14  20:00  Premier League  Matchweek 36  Fri  Away      W  4.0   
58 2021-05-18  19:00  Premier League  Matchweek 37  Tue  Away      L  2.0   
59 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  5.0   

     ga        opponent  ...   fk   pk  pkatt  season             team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022  Manchester City   
..  ...             ...  ...  ...  ...    ...     ...              ...   
54  0.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Manchester City   
56  2.0         Chelsea  ...  0.0  0.0    1.0    2021  Manchester City   
57  3.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Manchester City   
58  3.0        Brighton  ...  1.0  0.0    0.0    2021  Manchester City   
59  0.0         Everton  ...  0.0  0.0    0.0    2021  Manchester City   

   venue_code opponent_code hour  day_code  target  
1           0            18   16         6       0  
2           1            15   15         5       1  
3           1             0   12         5       1  
4           0            10   15         5       1  
6           1            17   15         5       0  
..        ...           ...  ...       ...     ...  
54          0             6   12         5       1  
56          1             5   17         5       0  
57          0            14   20         4       1  
58          0             3   19         1       0  
59          1             7   16         6       1  

[71 rows x 32 columns]

In [36]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values('date')
    rolling_stats = group[cols].rolling(3, closed='left').mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [37]:
cols = ['gf', 'ga', 'sh', 'sot', 'dist', 'fk', 'pk', 'pkatt']  # cols we want to compute rolling averages for
new_cols = [f'{c}_rolling' for c in cols]

In [38]:
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [39]:
rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  day_code  target  gf_rolling  ga_rolling  \
5   0.0         Arsenal  ...         5       1    2.000000    2.333333   
7   1.0        West Ham  ...         5       0    1.333333    2.000000   
9   0.0   Sheffield Utd  ...         5       1    1.000000    0.666667   
11  1.0       Liverpool  ...         6       0    1.000000    0.333333   
12  2.0       Tottenham  ...         5       0    1.000000    0.666667   
..  ...             ...  ...       ...     ...         ...         ...   
42  0.0  Crystal Palace  ...         0       0    2.333333    1.333333   
44  0.0         Burnley  ...         5       1    1.666667    0.333333   
46  2.0       Liverpool  ...         6       0    2.000000    0.333333   
49  0.0        Brighton  ...         2       1    1.333333    0.666667   
50  1.0         Watford  ...         5       1    2.333333    0.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5   17.333333    4.666667    18.900000   1.333333    0.333333       0.333333  
7   17.333333    3.666667    17.733333   0.666667    0.000000       0.000000  
9   16.666667    4.333333    18.233333   0.666667    0.000000       0.000000  
11  14.333333    6.666667    18.466667   1.000000    0.000000       0.000000  
12  12.000000    5.666667    19.366667   1.000000    0.000000       0.333333  
..        ...         ...          ...        ...         ...            ...  
42  19.000000    7.000000    15.366667   0.333333    0.333333       0.333333  
44  18.333333    7.333333    16.000000   0.333333    0.000000       0.000000  
46  20.000000    6.666667    16.133333   0.333333    0.000000       0.000000  
49  15.666667    4.666667    16.700000   0.333333    0.000000       0.000000  
50  15.333333    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

In [40]:
matches_rolling = matches.groupby('team').apply(lambda x: rolling_averages(x, cols, new_cols))

In [41]:
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling

In [42]:
matches_rolling = matches_rolling.droplevel('team') # we don't need team names

In [43]:
matches_rolling.head(5)

date   time            comp        round  day venue result   gf   ga  \
6  2020-10-04  14:00  Premier League  Matchweek 4  Sun  Home      W  2.0  1.0   
7  2020-10-17  17:30  Premier League  Matchweek 5  Sat  Away      L  0.0  1.0   
9  2020-10-25  19:15  Premier League  Matchweek 6  Sun  Home      L  0.0  1.0   
11 2020-11-01  16:30  Premier League  Matchweek 7  Sun  Away      W  1.0  0.0   
13 2020-11-08  19:15  Premier League  Matchweek 8  Sun  Home      L  0.0  3.0   

           opponent  ...  day_code  target  gf_rolling  ga_rolling sh_rolling  \
6     Sheffield Utd  ...         6       1    2.000000    1.333333   7.666667   
7   Manchester City  ...         5       0    1.666667    1.666667   5.333333   
9    Leicester City  ...         6       0    1.000000    1.666667   7.000000   
11   Manchester Utd  ...         6       1    0.666667    1.000000   9.666667   
13      Aston Villa  ...         6       0    0.333333    0.666667   9.666667   

   sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6     3.666667    14.733333   0.666667    0.000000       0.000000  
7     3.666667    15.766667   0.000000    0.000000       0.000000  
9     3.666667    16.733333   0.666667    0.000000       0.000000  
11    4.000000    16.033333   1.000000    0.000000       0.000000  
13    2.666667    18.033333   1.000000    0.333333       0.333333  

[5 rows x 40 columns]

In [44]:
matches_rolling.index = range(matches_rolling.shape[0])

In [45]:
matches_rolling.head(5)

date   time            comp        round  day venue result   gf   ga  \
0 2020-10-04  14:00  Premier League  Matchweek 4  Sun  Home      W  2.0  1.0   
1 2020-10-17  17:30  Premier League  Matchweek 5  Sat  Away      L  0.0  1.0   
2 2020-10-25  19:15  Premier League  Matchweek 6  Sun  Home      L  0.0  1.0   
3 2020-11-01  16:30  Premier League  Matchweek 7  Sun  Away      W  1.0  0.0   
4 2020-11-08  19:15  Premier League  Matchweek 8  Sun  Home      L  0.0  3.0   

          opponent  ...  day_code  target  gf_rolling  ga_rolling sh_rolling  \
0    Sheffield Utd  ...         6       1    2.000000    1.333333   7.666667   
1  Manchester City  ...         5       0    1.666667    1.666667   5.333333   
2   Leicester City  ...         6       0    1.000000    1.666667   7.000000   
3   Manchester Utd  ...         6       1    0.666667    1.000000   9.666667   
4      Aston Villa  ...         6       0    0.333333    0.666667   9.666667   

  sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0    3.666667    14.733333   0.666667    0.000000       0.000000  
1    3.666667    15.766667   0.000000    0.000000       0.000000  
2    3.666667    16.733333   0.666667    0.000000       0.000000  
3    4.000000    16.033333   1.000000    0.000000       0.000000  
4    2.666667    18.033333   1.000000    0.333333       0.333333  

[5 rows x 40 columns]

## Phase 5: Training the model again with new predictors

In [46]:
def make_predictions(data, predictors):
    train = data[data['date'] < '2022-01-01'] # take all matched before 2022
    test = data[data['date'] > '2022-01-01'] # take all matched before 2022
    rf.fit(train[predictors], train['target'])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test['target'], predicted=preds), index=test.index)
    precision = precision_score(test['target'], preds)
    return combined, precision

In [47]:
combined, precision = make_predictions(matches_rolling, predictors + new_cols)

In [48]:
precision # new precision

0.625

The result of the new model regarding changes made.

In [49]:
combined

actual  predicted
55         0          0
56         1          0
57         1          0
58         1          1
59         1          1
...      ...        ...
1312       1          0
1313       0          0
1314       1          0
1315       0          0
1316       0          0

[276 rows x 2 columns]

In [50]:
combined = combined.merge(matches_rolling[['date', 'team', 'opponent', 'result']], 
                                            left_index=True, right_index=True)

In [51]:
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

## Phase 6: Combining home and away predictions to improve predictions
It's worthy of mention that this part, can improve the result of the predictions without data manipulation!

In [52]:
class MissingDict(dict):
    __missing__ = lambda self, key: key
map_values = {
    'Brighton and Hove Albion': 'Brighton',
    'Manchester United': 'Manchester Utd',
    'Newcastle United': 'Newcastle Utd',
    'Tottenham Hotspur': 'Tottenham',
    'West Ham United': 'West Ham',
    'Wolverampton Wandrers': 'Wolves',


}
mapping = MissingDict(**map_values)

In [53]:
mapping['Arsenal']

'Arsenal'

In [54]:
mapping['West Ham United']

'West Ham'

In [55]:
combined['new_team'] = combined['team'].map(mapping)

In [56]:
combined.head(5)

actual  predicted       date     team   opponent result new_team
55       0          0 2022-01-23  Arsenal    Burnley      D  Arsenal
56       1          0 2022-02-10  Arsenal     Wolves      W  Arsenal
57       1          0 2022-02-19  Arsenal  Brentford      W  Arsenal
58       1          1 2022-02-24  Arsenal     Wolves      W  Arsenal
59       1          1 2022-03-06  Arsenal    Watford      W  Arsenal

Merging the df with itself

In [57]:
merged = combined.merge(combined, 
                        left_on=['date', 'new_team'], 
                        right_on=['date', 'opponent'])

In [58]:
merged

actual_x  predicted_x       date           team_x opponent_x result_x  \
0           0            0 2022-01-23          Arsenal    Burnley        D   
1           1            0 2022-02-10          Arsenal     Wolves        W   
2           1            0 2022-02-19          Arsenal  Brentford        W   
3           1            1 2022-02-24          Arsenal     Wolves        W   
4           1            1 2022-03-06          Arsenal    Watford        W   
..        ...          ...        ...              ...        ...      ...   
242         0            0 2022-03-20  West Ham United  Tottenham        L   
243         1            0 2022-04-03  West Ham United    Everton        W   
244         0            0 2022-04-10  West Ham United  Brentford        L   
245         0            0 2022-04-17  West Ham United    Burnley        D   
246         0            0 2022-04-24  West Ham United    Chelsea        L   

    new_team_x  actual_y  predicted_y                   team_y opponent_y  \
0      Arsenal         0            0                  Burnley    Arsenal   
1      Arsenal         0            0  Wolverhampton Wanderers    Arsenal   
2      Arsenal         0            0                Brentford    Arsenal   
3      Arsenal         0            0  Wolverhampton Wanderers    Arsenal   
4      Arsenal         0            0                  Watford    Arsenal   
..         ...       ...          ...                      ...        ...   
242   West Ham         1            1        Tottenham Hotspur   West Ham   
243   West Ham         0            0                  Everton   West Ham   
244   West Ham         1            1                Brentford   West Ham   
245   West Ham         0            0                  Burnley   West Ham   
246   West Ham         1            1                  Chelsea   West Ham   

    result_y               new_team_y  
0          D                  Burnley  
1          L  Wolverhampton Wanderers  
2          L                Brentford  
3          L  Wolverhampton Wanderers  
4          L                  Watford  
..       ...                      ...  
242        W                Tottenham  
243        L                  Everton  
244        W                Brentford  
245        D                  Burnley  
246        W                  Chelsea  

[247 rows x 13 columns]

In [59]:
merged[(merged['predicted_x'] == 1) & (merged['predicted_y'] == 0)]['actual_x'].value_counts()

1    26
0    13
Name: actual_x, dtype: int64

## Result
Calculating the final result of the prediction.

In [60]:
27 / 40

0.675